src/traces/treemap/attributes.js

```diff
@@ -0,0 +1,324 @@
+                'Note that the orientation of the generated rectangles (tall or wide)',
+                'is not implied by the ratio; for example, a ratio of two will attempt',
+                'to produce a mixture of rectangles whose width:height ratio is either 2:1 or 1:2.',
+                'When using *squarify*, unlike d3 which uses the Golden Ratio i.e. 1.618034,',
+                'Plotly applies 1 to increase squares in treemap layouts.'
+            ].join(' ')
+        },
+
+        mirror: {
+            valType: 'flaglist',
+            role: 'info',
+            flags: [
+                'x',
+                'y'
+            ],
+            dflt: '',
+            editType: 'plot',
+            description: [
+                'Determines if the positions obtained from solver are mirrored on each axis.'
+            ].join(' ')
+        },
+
+        pad: {
+            valType: 'number',
+            role: 'style',
+            min: 0,
+            dflt: 3,
+            editType: 'plot',
+            description: [
+                'Sets the inner padding (in px).'
+            ].join(' ')
+        },
+
+        editType: 'calc',
+    },
+
+    marker: extendFlat({
+        pad: {
+            top: {
+                valType: 'number',
+                role: 'style',
+                min: 0,
+                dflt: 'auto',
+                editType: 'plot',
+                description: [
+                    'Sets the padding form the top (in px).'
+                ].join(' ')
+            },
+            left: {
+                valType: 'number',
+                role: 'style',
+                min: 0,
+                dflt: 'auto',
+                editType: 'plot',
+                description: [
+                    'Sets the padding form the left (in px).'
+                ].join(' ')
+            },
+            right: {
+                valType: 'number',
+                role: 'style',
+                min: 0,
+                dflt: 'auto',
+                editType: 'plot',
+                description: [
+                    'Sets the padding form the right (in px).'
+                ].join(' ')
+            },
+            bottom: {
+                valType: 'number',
+                role: 'style',
+                min: 0,
+                dflt: 'auto',
+                editType: 'plot',
+                description: [
+                    'Sets the padding form the bottom (in px).'
+                ].join(' ')
+            },
+
+            editType: 'calc'
+        },
+
+        colors: {
+            valType: 'data_array',
+            editType: 'calc',
+            description: [
+                'Sets the color of each sector of this treemap chart.',
+                'If not specified, the default trace color set is used',
+                'to pick the sector colors.'
+            ].join(' ')
+        },
+
+        opacity: {
```

Comment:
Ok, thanks for the info.

Now I see that `marker.opacity` can multiply `branch.opacity`:

https://github.com/plotly/plotly.js/blob/eb54f2414b216dbd81d2fbd3f9525b70c74f5df8/src/traces/treemap/style.js#L43

That's interesting, but wouldn't having `marker.opacity` (when set) override `branch.opacity` be more user friendly in your mind? In particular, if we allow `marker.opacity` to be `arrayOk`, this could allow users to set the opacity of a sector pretty easily. 

What do you think @archmoj ?

Reply:
A very good call! 
It helped me rethink a couple of things in https://github.com/plotly/plotly.js/pull/4185/commits/0012fa1eaa8e57670df674a7b35d27413acb4d55. 
Now the API is more similar to `sunburst` where both have a `leaf.opacity` option.

Reply:
Hmm. Commit 0012fa1eaa8e57670df674a7b35d27413acb4d55 looks a little odd to me.

Having `marker.opacity` now essentially be the base of the `base^pt.height` computation may be even less user-friendly than what you had before. Moreover, I can't imagine users setting `leaf.opacity` to anything other than `1` in treemap traces (I could be wrong here).

Granted, maybe exposing the `base` parameter in `base^pt.height` sector computation as an attribute is the way to go, but in this case, it should **not** be called `marker.opacity`. Something like `marker.opacitybase` would be much better.

@archmoj does that make sense?

Reply:
Conclusion:

- drop `leaf.opacity`
- rename `marker.opacity` -> `marker.opacitybase` (or something better, but not `marker.opacity`)

Reply:
`marker.depthfade`?

It also occurs to me that opacity might not actually be quite the effect we want - perhaps it would be better to use the opaque color that would result from blending the marker color with the background color in the same proportion? For the bottom-most layer the effect is equivalent, but anything stacked above it currently effectively is more opaque (if the two colors are the same) or blended to a different color (if using a colorscale or explicit colors). Both of these I think are problematic:
- if the color has meaning, a faded version of that color is probably still easy to mentally connect with the original, keeping its interpretation the same, but a version blended with a different color could shift the implied meaning of that color.
- making stacked layers more opaque kind of defeats the purpose - it means that the final effective opacity drops off too slowly at first, then more quickly toward the back of the stack.

Anyway if we extend the analogy I mentioned this morning of "looking at layers of hills in the distance" - those hills are not transparent, they are faded by the air in front of them - which is exactly the same effect as blending with the background color.

Reply:
@alexcjohnson Thanks very much for the feedback. Is there a place in plotly.js code that a similar colour blending is applied?

Reply:
look for color.combine

https://github.com/plotly/plotly.js/blob/56f6983360febe9d5b3bf7427ada9863c372561a/src/components/color/index.js#L46-L62

Reply:
Awesome!

Reply:
New attempt in https://github.com/plotly/plotly.js/pull/4185/commits/4c7a092a385f6db7e3ed85c59d5ba1f344446650 - which I'm a fan of :ok_hand: